### DISCUSSION TITLE
Setting and/or Modifying  PageMode key in document catalog

### DISCUSSION THREAD

**Discussion by PeterSlezak** [pymupdf/PyMuPDF#2002] [Q&A] [answered]
Hi,
I’m combining several pdf into one file and inserting outlines using set_toc afterwards to navigate to the beginning of each file within output pdf. Everything works just fine; I just want to set the /PageMode /UseOutlines to the final PDF so that outlines navigation panel is visible when pdf file is open in pdf viewer. Adobe acrobat call this initial view, navigation tab property as “Bookmarks Panel and Page”. 
What I currently do is to create a new pdf document using:  
```
def NewOpen():
    '''update root of new pdf file to display bookmarks by default'''
    pdffile = fitz.open()
    cat = pdffile.pdf_catalog()
    li = pdffile.xref_object(cat).rsplit('>>', 1)
    updt = '/PageMode /UseOutlines >>'.join(li)
    pdffile.update_object(cat, updt)
    return pdffile
```
and inserting individual input PDF into it afterwards.
Is there an easier way to achieve this? Alternatively, would it be possible to add parameter to fitz.open() (or to set_toc itself) to set this /PageMode key and value to Document Catalog object?
Thanks!

**Comment by JorjMcKie** [comment, marked as answer]
> Is there an easier way to achieve this? Alternatively, would it be possible to add parameter to fitz.open() (or to set_toc itself) to set this /PageMode key and value to Document Catalog object?

There are two questions:

1. Is there an easier way to achieve this?
    
    - Yes, there is: you can set PDF dictionary key-value pairs with method `doc.xref_set_key(doc.pdf_catalog(), "PageMode", "/UseOutlines")`. This does what you already have done so marvelously using string manipulation. The advantages of the method are that it (1) automatically resolves multiple nested xref sub-references, (2) automatically inserts, resp. overwrites the key (here `/PageMode`) if not present, respectively already there. If you use the special PDF name for the `None` equivalent "null", a key can also be removed from the containing parent. Plus, you can use a hierarchy of keys like `doc.xref_set_key(xref, "key1/key2/...", "value")`. Note that the first slash must be omitted.

2. In light of the previous relatively simple one-liner (for people with PDF expertise like you) and in view of the dozens of other global properties definable in the PDF catalog, we are currently not planning to create special purpose functions for this.

**Reply by PeterSlezak** [reply]
Thank! That's perfect, I'll definitely use the xref_set_key a lot from now on.

**Reply by JorjMcKie** [reply]
Glad about that. Best you look up the docu, because a lot more has to be said about it.
